### PR TITLE
Extracts empty type as an abstract (empty-by-construction) type in OCaml

### PR DESCRIPTION
**Pull request by herbelin**
**Kind:** enhancement

It is unclear why the empty type is extracted as the unit type in extraction to OCaml. This PR extracts the empty type to the empty type.

Addresses a remark at [#14801](https://github.com/coq/coq/issues/14801#issuecomment-903827850).

- [x] Added / updated test-suite
- [x] Entry added in the changelog

Would need to be done in Haskell too.

### CODE REVIEW

**Comment by ppedrot** [comment]
Probably historical reasons, there used not to be an empty type in OCaml not so long ago.

File: plugins/extraction/ocaml.ml [hunk 1]
```diff
@@ -494,9 +494,9 @@ let pp_one_ind prefix ip_equiv pl name cnames ctyps =
             (fun () -> spc () ++ str "* ") (pp_type true pl) typs)
   in
   pp_parameters pl ++ str prefix ++ name ++
-  pp_equiv pl name ip_equiv ++ str " =" ++
-  if Int.equal (Array.length ctyps) 0 then str " unit (* empty inductive *)"
-  else fnl () ++ v 0 (prvecti pp_constructor ctyps)
+  pp_equiv pl name ip_equiv ++
+  if Int.equal (Array.length ctyps) 0 then mt ()
```
Comment (ppedrot) [inline plugins/extraction/ocaml.ml:498]: This is not the empty type, this is merely an abstract type. You should write `= |` as [per the doc](https://ocaml.org/manual/emptyvariants.html#s:empty-variants).

**Comment by SkySkimmer** [comment]
This seems to extract to an abstract type, not an empty type (`type empty = |` since ocaml 4.07.0, otherwise `type empty = { any : 'a. 'a }` or `type (_,_) empty_aux = Refl : ('a,'a) empty_aux;; type empty = (unit,bool) empty_aux` or some other method)

**Comment by herbelin** [comment]
How do these variants observationally differ compared to an abstract type (in the implementation)?

**Comment by ppedrot** [comment]
You can pattern-match an empty type, and flambda probably takes advantage of this knowledge to further optimize the code.

**Comment by herbelin** [comment]
Extraction seems to discard `match` on an empty type anyway...

On another side, is OCaml 4.07 (Summer 2018) a reasonable requirement for extraction? (It would only be to support empty types anyway.)

**Comment by Zimmi48** [comment]
> On another side, is OCaml 4.07 (Summer 2018) a reasonable requirement for extraction? (It would only be to support empty types anyway.)

I believe everyone agreed it would even be a reasonable dependency for Coq.

**Comment by herbelin** [comment]
I then moved to `= |`. The situation is probably not so common and we can hope that the 4.07.0 requirement will not be a handicap.

**Comment by Alizter** [comment]
Do we have tests for extractions? I wonder if adding at test like checking the type of `¬¬¬A -> ¬A` might be worth it?

**Comment by herbelin** [comment]
> Do we have tests for extractions?

There are some in `test-suite/success` and `test-suite/output`.

> I wonder if adding at test like checking the type of ¬¬¬A -> ¬A extract in ocaml might be worth it?

Why not adding a test. I don't know exactly what you have in mind with ¬¬¬A -> ¬A but if ever you have a test ready, don't hesitate to push it to this PR or to add it in a comment.

**Comment by herbelin** [comment]
About the CI, bad luck, metacoq does not use a version of OCaml recent enough to support `type foo = |`.

To metacoq maintainers: would you prefer to upgrade the OCaml requirements or do I backtrack on using an abstract type, which I believe would work essentially the same wrt extraction (since `match` on empty types are dropped).

**Comment by Alizter** [comment]
@herbelin I've added a test. Do you want to see if using an abstract type does the same thing?

**Comment by herbelin** [comment]
> @herbelin I've added a test.

Nice!

> Do you want to see if using an abstract type does the same thing?

Do you mean changing the test so that it assumes instead extraction to an abstract type? (Or do you mean the PR to be changed back to using an empty type so as to comply with the test?)

**Comment by Alizter** [comment]
Hmm, actually I'm not sure what I meant. What needs to be done for this PR now?

**Comment by herbelin** [comment]
I got confused. The PR is currently using an empty type but this breaks metacoq as well as the new test due to OCaml < 4.07 being used.

I think it is safe to use an abstract type. So, I would suggest to forget about waiting about the opinion of metacoq and to go back to an abstract type (and not the converse as my previous comment wrongly said). What do you think?

**Comment by Alizter** [comment]
I'm fine with either empty type. I don't think the test will act too differently, but we should definitely inspect if the expected output is correct.

**Review by Alizter — APPROVED** [review]

**Comment by Alizter** [comment]
Needs an assignee cc @ppedrot ?

**Comment by ppedrot** [comment]
I am not an extraction maintainer but that will do it...

**Comment by ppedrot** [comment]
What about #14802?

**Comment by Alizter** [comment]
@ppedrot Did you mean #9416?

**Comment by silene** [comment]
> I believe everyone agreed it would even be a reasonable dependency for Coq.

That is not how I remember it. I think the consensus was that the language did not evolve sufficiently between 4.05 and 4.07 to warrant bumping the minimum version of the compiler. OCaml 4.08, however, started providing monadic lets, and it was deemed a feature worth bumping the minimum version for. Unfortunately, Coq, when compiled with OCaml 4.08, has some known performance issues (?). So, we would need to bump the minimum version even further. But we could not agree on the exact version.

That said, this is kind of orthogonal, and we could imagine that extraction requires a more recent version of OCaml than Coq itself.

**Review by ppedrot — DISMISSED** [review]
This needs a changelog.

**Comment by herbelin** [comment]
> This needs a changelog.

Added.

> Would need to be done in Haskell too.

There are different ways to do an empty type in Haskell (see e.g. [here](https://wiki.haskell.org/Empty_type)), but it is not obvious to me what kind of compatibility incantation we would have to give (such as enabling the [`EmptyDataDecls`](https://ghc.gitlab.haskell.org/ghc/doc/users_guide/exts/nullary_types.html) GHC). Leaving it to the specialists.

**Review by ppedrot — APPROVED** [review]

**Pull request merged**